src/goto-instrument/code_contracts.cpp

```diff
@@ -961,3 +940,713 @@ bool code_contractst::enforce_contracts(
+    init_block.add(goto_programt::make_assignment(
+      code_assignt(upper_offset_variable, upper_offset_object),
+      function_symbol.location));
+  }
+  else
+  {
+    exprt upper_constant_size =
+      get_size(array.type().subtype(), contract.ns, log);
+    upper_offset_object = typecast_exprt(
+      mult_exprt(
+        typecast_exprt(upper_operand, unsigned_long_int_type()),
+        upper_constant_size),
+      signed_int_type());
+
+    // Declare a new symbol to stand in for the reference
+    symbolt upper_standin_symbol = contract.new_tmp_symbol(
+      upper_offset_object.type(),
+      function_symbol.location,
+      function_id,
+      function_symbol.mode);
+
+    upper_offset_variable = upper_standin_symbol.symbol_expr();
+
+    // Add array temp to variable initialization block
+    init_block.add(goto_programt::make_decl(
+      upper_offset_variable, function_symbol.location));
+    init_block.add(goto_programt::make_assignment(
+      code_assignt(upper_offset_variable, upper_offset_object),
+      function_symbol.location));
+  }
+}
+
+std::vector<symbol_exprt>
+assigns_clause_array_targett::temporary_declarations() const
+{
+  std::vector<symbol_exprt> result;
+  result.push_back(array_standin_variable);
+  result.push_back(lower_offset_variable);
+  result.push_back(upper_offset_variable);
+
+  return result;
+}
+
+goto_programt
+assigns_clause_array_targett::havoc_code(source_locationt location) const
+{
+  goto_programt assigns_havoc;
+
+  modifiest assigns_tgts;
+  typet lower_type = lower_offset_variable.type();
+  exprt array_type_size =
+    get_size(pointer_object.type().subtype(), contract.ns, log);
+
+  for(mp_integer i = lower_bound; i <= upper_bound; ++i)
+  {
+    irep_idt offset_string(from_integer(i, integer_typet()).get_value());
+    irep_idt offset_irep(offset_string);
+    constant_exprt val_const(offset_irep, lower_type);
+    dereference_exprt array_deref(plus_exprt(
+      pointer_object, typecast_exprt(val_const, signed_long_int_type())));
+
+    assigns_tgts.insert(array_deref);
+  }
+
+  for(auto lhs : assigns_tgts)
+  {
+    side_effect_expr_nondett rhs(lhs.type(), location);
+
+    goto_programt::targett target =
+      assigns_havoc.add(goto_programt::make_assignment(
+        code_assignt(std::move(lhs), std::move(rhs)), location));
+    target->code_nonconst().add_source_location() = location;
+  }
+
+  return assigns_havoc;
+}
+
+exprt assigns_clause_array_targett::alias_expression(const exprt &ptr)
+{
+  exprt same_obj = same_object(ptr, array_standin_variable);
+  exprt ptr_offset = pointer_offset(ptr);
+
+  exprt in_range_lower = binary_predicate_exprt(
+    ptr_offset,
+    ID_ge,
+    typecast_exprt(lower_offset_variable, ptr_offset.type()));
+  exprt in_range_upper = binary_predicate_exprt(
+    typecast_exprt(upper_offset_variable, ptr_offset.type()),
+    ID_ge,
+    ptr_offset);
+  exprt in_range = and_exprt(in_range_lower, in_range_upper);
+
+  // Troublesome
```

Comment:
What is troublesome here? And why? Does it need to be guarded by an Invariant maybe?

Reply:
Fixed it.